### PR TITLE
Allocate an auxiliary struct when running out of preconfigured threads

### LINKED ISSUE #1141: single thread cause core dump
I make the source code to generate the libopenblas*.a, and use the APIs in my.a. Then I make a new.so  including both libopenblas*.a and my.a,  and make an executable file that reference this new.so to test my code. When I use the multi-thread libopenblas*.a(just use "make ONLY_CBLAS=1"), the main() function works normally. but when I use the single-thread libopenblas*.a(use "make USE_THREAD=0"), the main() causes core dump--"segmentation fault(core dump)".
    I tried version 0.2.18 and 0.2.20, but both has the core dump, even I just called functions that donot call openblas APIs.
    My platform is allwinner-R16, CPU is ARMV7, and I made the libopenblas*.a with cross compiler.

### LINKED ISSUE #2225: OpenBLAS crashing for Julia with different threading options
Hi, I've posted this to the Julia issues but figured I would link here as well

https://github.com/JuliaLang/julia/issues/14857#issuecomment-522257442

My machine has 56 cores, and the default for OPENBLAS_NUM_THREADS on my machine seems to be 8- I don't remember if I had any control in that, as I don't believe I built OpenBLAS from source

Running with half the julia threads works for me:
```
$ OPENBLAS_NUM_THREADS=8 JULIA_NUM_THREADS=28 julia13 --start=no
julia> versioninfo()
Julia Version 1.3.0-alpha.0
Commit 6c11e7c2c4 (2019-07-23 01:46 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, broadwell)
Environment:
  JULIA_NUM_THREADS = 28

julia> ccall((:openblas_get_num_threads64_, Base.libblas_name), Cint, ()), Threads.nthreads()
(8, 28)

julia> include("./blas_thread_test.jl")
190.196333 seconds (15.18 M allocations: 384.406 GiB, 9.51% gc time)
```

But other settings segfault with the same error `BLAS : Program is Terminated. Because you tried to allocate too many memory regions.

```
$ OPENBLAS_NUM_THREADS=8 JULIA_NUM_THREADS=56 julia13 --start=no

julia> ccall((:openblas_get_num_threads64_, Base.libblas_name), Cint, ()), Threads.nthreads()
(8, 56)

julia> include("./blas_thread_test.jl")
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
....
signal (11): Segmentation fault
in expression starting at /home/scott/repos/blas_thread_test.jl:23
....
(let me know if you want the full long stack trace)
```

```
$ OPENBLAS_NUM_THREADS=1 JULIA_NUM_THREADS=56 julia13 --start=no

julia> ccall((:openblas_get_num_threads64_, Base.libblas_name), Cint, ()), Threads.nthreads()
(1, 56)

julia> include("./blas_thread_test.jl")
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.

signal (11): Segmentation fault
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
in expression starting at /home/sn expression starting at /home/scott/repos/blas_thread_test.jl:23
```

Here's the test program I'm running:
```
$ cat blas_thread_test.jl
using LinearAlgebra
function irls(A, b; iters=100)
    M, N = size(A)
    x = zeros(eltype(b), N)
    ep = sqrt(eps(eltype(A)))
    p = 1

    W = diagm(0 => (abs.(b-A*x) .+ ep).^(p-2))

    for ii in 1:iters
        x .= (A' * W * A) \ (A' * W * b)
        W .= diagm(0 => (abs.(b-A*x) .+ ep).^(p-2))
    end
    return x
end


M, N = 500, 30
A = randn(Float32, M, N);
bstack = rand(Float32, 60, 60, M);
xs = zeros(Float32, 60, 60, N);

@time Threads.@threads for j=1:size(bstack, 2)
    for i = 1:size(bstack, 1)
        xs[i, j, :] .= irls(A, bstack[i, j, :], iters=100)
    end
end
```

### LINKED ISSUE #2304: using large NUM_THREADS for single-threaded openblas
Hi,

I am using single-threaded openblas on armv8 (built with "make TARGET=ARMV8 USE_THREAD=0 USE_LOCKING=1 CC=aarch64-linux-gnu-gcc FC=aarch64-linux-gnu-gfortran HOSTCC=gcc"). In my program, sgemm is called in parallel, and when number of threads exceeds NUM_BUFFERS, error "BLAS : Program is Terminated. Because you tried to allocate too many memory regions." will occur.

Then I'm wondering if I could set NUM_THREADS to a large value (like 10,000) so that NUM_BUFFERS is definitely greater than number of threads. I want to know how much memory is needed per buffer, and is it safe to set a large NUM_THREADS.

### LINKED ISSUE #3321: BLAS memory allocation error in Scikit-learn KMeans & kNN & DBSCAN
[https://github.com/scikit-learn/scikit-learn/issues/20539](https://github.com/scikit-learn/scikit-learn/issues/20539)

Do you have any ideas?

### PR DESCRIPTION
**Pull request by martin-frbg**
for #3321

### CODE REVIEW

**Comment by mattip** [comment]
Trying this out in MacPython/openblas-libs#66

File: driver/others/memory.c [hunk 1]
```diff
@@ -73,6 +73,16 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include "common.h"
 
+#ifndef likely
+#ifdef __GNUC__
```
Comment (isuruf) [inline driver/others/memory.c:77]: How about `#if defined(__GNUC__) || defined(__clang__)` ?

Comment (martin-frbg) [reply driver/others/memory.c:77]: is clang likely to understand this (across all relevant versions) ? I copied this from the `blas_server*` files...

Comment (isuruf) [reply driver/others/memory.c:77]: clang 3.8 does which was released 5 years ago. That's the oldest I have access to. Here's what boost does, https://github.com/boostorg/config/blob/develop/include/boost/config/compiler/clang.hpp#L71-L77

**Pull request merged**